### DISCUSSION TITLE
MongoDB as replicat set

### DISCUSSION THREAD

**Discussion by unpseudocomplique** [coollabsio/coolify#1716] [Feature Requests]
Hi, 
i want to have my mongodb as replicat set.

There is multiples reasons. 
First always good to have multiples servers.
But main reason rn is to be able to have access to change streams. I'm on 6.0.11 and i need to be able to have a change stream event for things like analytics, webhook, notification, calculated properties... There is a lot we want to to with change stream!

**Comment by pranavmappoli** [comment]
@unpseudocomplique Did you find any way to achieve this ? Prisma wont work without replica set :(

**Comment by unpseudocomplique** [comment]
@pranavmappoli nope, i need replicat set for change stream and perf but i'm using mongoose so it's not blocking for now

**Comment by rufat** [comment]
I will explain how I set up a MongoDB replica on Coolify.

1. You will need to create three MongoDB instances on Coolify.
<img width="1179" alt="image" src="https://github.com/coollabsio/coolify/assets/5330471/997cd352-78d8-4fc5-a609-fa608705ac49">

2.  Give unique ports such as **5432, 5433, 5434** and turn on **Make it publicly available** on each of the instances.

3. Open terminal via ssh and create a key by using: `openssl rand -base64 741`.

4. Copy the output and create a key file named **keyfile-mongo-0** and save it.

5. Give reading permission to it by using **chmod 400 keyfile-mongo-0**.

6. Go to Coolify and add this to the **Custom MongoDB Configuration** textarea:
```
replication:
  replSetName: "rs0"
security:
  authorization: enabled
  keyFile: /tmp/root/keyfile-mongo-0
```
7. Go to Storages on Coolify and create a volume. **/root/mongo-repl-0-mount** is where your key file is located at your host server.
<img width="776" alt="image" src="https://github.com/coollabsio/coolify/assets/5330471/bc83d1a8-7895-4971-acac-066f7722d400">

8. Finally, connect first MongoDB instace and run this mongosh command:
```
rs.initiate({
 _id: "rs0",
 members: [
   {_id: 0, host: "<your_server_ip>:5432"},
   {_id: 1, host: "<your_server_ip>:5433"},
   {_id: 2, host: "<your_server_ip>:5434"}
 ]
})
```

This is it. Connect to your replicate set with this format of connection string:
`mongodb://<username>:<password>@<your_server_ip>:5432,<your_server_ip>:5433,<your_server_ip>:5434/?replicaSet=rs0`

**Reply by toverux** [reply]
One thing to add, I had to do a `chown 999:999 mongo-repl-0-mount/keyfile-mongo-0`, 999 being the mongod user in the container. Otherwise, it can't read the file (the error log is explicit about that).

I also had an issue after executing `rs.initiate()` (without args), my only replica set member (*you don't have to create three like in this example*) was not added correctly to the replica set, and as such MongoDB wasn't able to elect a Primary. This solved it for me: https://dba.stackexchange.com/a/299546 (replacing the "random" host with just localhost)

It would be nice that Coolify supported replica sets. First, the non-replica set mode of Mongo is kinda legacy-ish, and second a lot of clients/ORMs/apps/etc require replica sets now, because this enables some useful Mongo features like transactions.

**Reply by radud** [reply]
hi there, 

In my case nodes cannot access each other if they are referred by external ip. I can access them fine from outside but for some reason they cannot connect to each other. They can connect only if they are referred by docker network name like:

rs.initiate({
 _id: "rs0",
 members: [
   {_id: 0, host: "ksccogogo8o8w8oc8cokk8gs:27017"},
   {_id: 1, host: "d0k4oswkkcgw4cok0socoo8o:27017"},
   {_id: 2, host: "vcos8s0s0w8wg4sw48gk0k4o:27017"}
 ]
})


Anyone encountered this?

**Reply by theDanielJLewis** [reply]
@radud, here's how I figured out how to do it. And from the MongoDB documentation and forums, it seems this is the only way.

Use subdomains.

But here's the key: edit the `/etc/hosts` file _on each server_ to make each domain point to `127.0.0.1` (or the private network address, like 10.0.0.x).

For example:

```
127.0.0.1 mongo1.mydomain.com
127.0.0.1 mongo2.mydomain.com
127.0.0.1 mongo3.mydomain.com
```

But in your domain DNS records, ensure those subdomains point to the public IPv4 address(es).

Then in your initiate command, use those subdomains instead of IP addresses or the Docker network names. You'll also use the subdomains in your connections from your app or `mongosh`.

And don't forget to open whatever ports you assigned for each MongoDB instance!

**Reply by cbratschi** [reply]
> One thing to add, I had to do a `chown 999:999 mongo-repl-0-mount/keyfile-mongo-0`, 999 being the mongod user in the container. Otherwise, it can't read the file (the error log is explicit about that).

I could not get past "Unable to acquire security key[s]". Tried to mount a volume or entered the keyfile data as a file then used chown to change the user but this did not help.

**Reply by cbratschi** [reply]
In the destination path I had to add the full file name and not only the path e.g. /tmp/root/mongodb-rs0-keyfile. After that the file was available and the permission were correct.
